src/mlpack/methods/ann/layer/group_norm_impl.hpp

```diff
@@ -0,0 +1,186 @@
+
+namespace mlpack {
+namespace ann { /** Artificial Neural Network. */
+
+
+template<typename InputDataType, typename OutputDataType>
+GroupNorm<InputDataType, OutputDataType>::GroupNorm() :
+    channelSize(0),
+    size(0),
+    groupCount(1),
+    eps(1e-8),
+    loading(false)
+{
+  // Nothing to do here.
+}
+
+template <typename InputDataType, typename OutputDataType>
+GroupNorm<InputDataType, OutputDataType>::GroupNorm(
+    const size_t channelSize, const size_t size, const size_t groupCount, const double eps) :
+    channelSize(channelSize),
+    size(size),
+    groupCount(groupCount),
+    eps(eps),
+    loading(false)
+{
+  if (size % groupCount != 0)
+  {
+    Log::Fatal << "Total input units must be divisible by groupCount!" << std::endl;
+  }
+
+  weights.set_size(size * 2 + 1, 1);
+}
+
+template<typename InputDataType, typename OutputDataType>
+void GroupNorm<InputDataType, OutputDataType>::Reset()
+{
+  gamma = arma::mat(weights.memptr(), size, 1, false, false);
+  beta = arma::mat(weights.memptr() + gamma.n_elem, size, 1, false, false);
+
+  if (!loading)
+  {
+    gamma.fill(1.0);
+    beta.fill(0.0);
+  }
+
+  loading = false;
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void GroupNorm<InputDataType, OutputDataType>::Forward(
+    const arma::Mat<eT>& input, arma::Mat<eT>& output)
+{
+  assert(input.n_rows % channelSize == 0);
+  assert(input.n_rows % size == 0);
+  arma::mat reshapedInput(const_cast<arma::Mat<eT>&>(input).memptr(),
+    input.n_rows * groupCount / size, input.n_cols * size / groupCount, false, false);
+
+  if (output.is_empty())
+    output.zeros(input.n_rows, input.n_cols);
+
+  arma::mat reshapedOutput((output).memptr(),
+    input.n_rows * groupCount / size, input.n_cols * size / groupCount, false, false);
+
+  mean = arma::mean(reshapedInput, 0);
+  variance = arma::var(reshapedInput, 1, 0);
+
+  // Normalize the input.
+  reshapedOutput += reshapedInput.each_row() - mean;
+  inputMean = reshapedOutput;
+  reshapedOutput.each_row() /= arma::sqrt(variance + eps);
+
+  // Reused in the backward and gradient step.
+  normalized = output;
+
+  arma::mat expandedGamma, expandedBeta;
+  expandedGamma.set_size(output.n_rows, 1);
+  expandedBeta.set_size(output.n_rows, 1);
+  for (size_t r = 0; r < output.n_rows; ++r)
+  {
+    expandedGamma(r) = gamma(r % size);
+    expandedBeta(r) = beta(r % size);
+  }
+
+  // Scale and shift the output.
+  output.each_col() %= expandedGamma;
+  output.each_col() += expandedBeta;
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void GroupNorm<InputDataType, OutputDataType>::Backward(
+    const arma::Mat<eT>& input, const arma::Mat<eT>& gy, arma::Mat<eT>& g)
+{
+  arma::mat inputReshaped(const_cast<arma::Mat<eT>&>(input).memptr(),
+    size / groupCount, groupCount * input.n_cols, false, false);
+  arma::mat gReshaped((g).memptr(),
+    size / groupCount, groupCount * g.n_cols, false, false);
```

Comment:
```suggestion
      size / groupCount, groupCount * input.n_cols, false, false);
  arma::mat gReshaped((g).memptr(),
      size / groupCount, groupCount * g.n_cols, false, false);
```

Should be tabbed twice (4 spaces).